v1 and v2 specs (#276)
zarr-developers/zarr-specs: docs/v2/v2.0.rst

    produce a directory-like structure.

Other keys SHOULD NOT be present within the metadata object and SHOULD be
ignored by implementations.

For example, the JSON object below defines a 2-dimensional array of 64-bit
little-endian floating point numbers with 10000 rows and 10000 columns, divided
into chunks of 1000 rows and 1000 columns (so there will be 100 chunks in total
arranged in a 10 by 10 grid). Within each chunk the data are laid out in C
contiguous order. Each chunk is encoded using a delta filter and compressed
using the Blosc compression library prior to storage::

    {
        "chunks": [
            1000,
            1000
        ],
        "compressor": {
            "id": "blosc",
            "cname": "lz4",
            "clevel": 5,
            "shuffle": 1
        },
        "dtype": "<f8",
        "fill_value": "NaN",
        "filters": [
            {"id": "delta", "dtype": "<f8", "astype": "<f4"}
        ],
        "order": "C",
        "shape": [
            10000,
            10000
        ],
        "zarr_format": 2
    }

.. _spec_v2_array_dtype:

Data type encoding
~~~~~~~~~~~~~~~~~~

Simple data types are encoded within the array metadata as a string,

> following the :ref:`NumPy array protocol type string (typestr) format
> <numpy:arrays.interface>`. The format

Reviewer: ```suggestion
following the `NumPy array protocol type string (typestr) format <https://numpy.org/doc/stable/reference/arrays.interface.html>`_. The format
```
Fixed broken link.

consists of 3 parts:

* One character describing the byteorder of the data (``"<"``: little-endian;
  ``">"``: big-endian; ``"|"``: not-relevant)
* One character code giving the basic type of the array (``"b"``: Boolean (integer
  type where all values are only True or False); ``"i"``: integer; ``"u"``: unsigned
  integer; ``"f"``: floating point; ``"c"``: complex floating point; ``"m"``: timedelta;
  ``"M"``: datetime; ``"S"``: string (fixed-length sequence of char); ``"U"``: unicode
  (fixed-length sequence of Py_UNICODE); ``"V"``: other (void * – each item is a
  fixed-size chunk of memory))
* An integer specifying the number of bytes the type uses.

The byte order MUST be specified. E.g., ``"<f8"``, ``">i4"``, ``"|b1"`` and
``"|S12"`` are valid data type encodings.

For datetime64 ("M") and timedelta64 ("m") data types, these MUST also include the
units within square brackets. A list of valid units and their definitions are given in

> the :ref:`NumPy documentation on Datetimes and Timedeltas
> <numpy:arrays.dtypes.dateunits>`.

Reviewer: ```suggestion
the `NumPy documentation on Datetimes and Timedeltas <https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units>`_.
```

For example, ``"<M8[ns]"`` specifies a datetime64 data type with nanosecond time units.

Structured data types (i.e., with multiple named fields) are encoded

> as a list of lists, following :ref:`NumPy array protocol type descriptions
> (descr)
> <numpy:arrays.interface>`. Each

Reviewer: ```suggestion
as a list of lists, following `NumPy array protocol type descriptions
(descr) <https://numpy.org/doc/stable/reference/arrays.interface.html#the-array-interface-protocol>`_. Each
```

    >>> sorted(os.listdir('data/group.zarr/foo/bar'))
    ['.zarray', '.zattrs', '0.0', '0.1', '1.0', '1.1']

Here is the same example using a Zip file as storage::

    >>> store = zarr.ZipStore('data/group.zip', mode='w')
    >>> root_grp = zarr.group(store)
    >>> sub_grp = root_grp.create_group('foo')
    >>> a = sub_grp.create_dataset('bar', shape=(20, 20), chunks=(10, 10))
    >>> a[:] = 42
    >>> a.attrs['comment'] = 'answer to life, the universe and everything'
    >>> store.close()

What has been stored::

    >>> import zipfile
    >>> zf = zipfile.ZipFile('data/group.zip', mode='r')
    >>> for name in sorted(zf.namelist()):
    ...     print(name)
    .zgroup
    foo/.zgroup
    foo/bar/.zarray
    foo/bar/.zattrs
    foo/bar/0.0
    foo/bar/0.1
    foo/bar/1.0
    foo/bar/1.1

.. _spec_v2_changes:

Changes
-------

Version 2 clarifications
~~~~~~~~~~~~~~~~~~~~~~~~

The following changes have been made to the version 2 specification since it was
initially published to clarify ambiguities and add some missing information.

* The specification now describes how bytes fill values should be encoded and
  decoded for arrays with a fixed-length byte string data type (`#165

>   <https://github.com/zarr-developers/zarr-python/issue/165/>`_, `#176

Reviewer: ```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/165/>`_, `#176
```

>   <https://github.com/zarr-developers/zarr-python/issue/176/>`_).

Reviewer: ```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/176/>`_).
```

* The specification now clarifies that units must be specified for datetime64 and
  timedelta64 data types (`#85

>   <https://github.com/zarr-developers/zarr-python/issue/85/>`_, `#215

Reviewer: ```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/85/>`_, `#215
```

>   <https://github.com/zarr-developers/zarr-python/issue/215/>`_).

Reviewer: ```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/215/>`_).
```

* The specification now clarifies that the '.zattrs' key does not have to be present for
  either arrays or groups, and if absent then custom attributes should be treated as
  empty.

* The specification now describes how structured datatypes with
  subarray shapes and/or with nested structured data types are encoded
  in array metadata (`#111

>   <https://github.com/zarr-developers/zarr-python/issue/111/>`_, `#296

Reviewer: ```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/111/>`_, `#296
```

>   <https://github.com/zarr-developers/zarr-python/issue/296/>`_).

Reviewer: ```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/296/>`_).
```
